## Chapter 4 – Exercise 3:  Sử dụng Mysql Connector để thực hiện các truy vấn dữ liệu trên nhiều bảng

#### Học viên thực hiện các yêu cầu sau và đối chiếu với kết quả cho trước:

In [1]:
from mysql.connector import connect

In [10]:
conn = connect(host='localhost', port=3306, database='salesdb', user='root', password='')

In [14]:
def run_sql(sql):
    # code here
    curs = conn.cursor()
    curs.execute(sql)
    results = curs.fetchall()
    curs.close()
    return results
    
    

In [16]:
# Câu 1: Liệt kê danh sách sản phẩm theo từng nhà cung cấp, thông tin gồm ProductId, ProductName, UnitPrice, CompanyName
sql_cmd1 = '''
SELECT p.Id as ProductId, p.ProductName, p.UnitPrice, s.CompanyName
FROM `products` p JOIN `suppliers` s ON p.SupplierId = s.Id;
'''

results_1 = run_sql(sql_cmd1)

for row in results_1:
    print(row)

(1, 'Chai', Decimal('18.00'), 'Exotic Liquids')
(2, 'Chang', Decimal('20.00'), 'Exotic Liquids')
(3, 'Aniseed Syrup', Decimal('10.00'), 'Exotic Liquids')
(4, "Chef Anton's Ca6/Seasoning", Decimal('22.00'), 'New Orleans Ca6/Delights')
(5, "Chef Anton's Gumbo Mix", Decimal('21.35'), 'New Orleans Ca6/Delights')
(6, "Grandma's Boysenberry Spread", Decimal('25.00'), "Grandma Kelly's Homestead")
(7, "Uncle Bob's Organic Dried Pears", Decimal('30.00'), "Grandma Kelly's Homestead")
(8, 'Northwoods Cranberry Sauce', Decimal('40.00'), "Grandma Kelly's Homestead")
(9, 'Mishi Kobe Niku', Decimal('97.00'), 'Tokyo Traders')
(10, 'Ikura', Decimal('31.50'), 'Tokyo Traders')
(11, 'Queso Cabrales', Decimal('21.00'), "Cooperativa de Quesos 'Las Cabras'")
(12, 'Queso Manchego La Pastora', Decimal('38.00'), "Cooperativa de Quesos 'Las Cabras'")
(13, 'Konbu', Decimal('6.00'), "Mayumi's")
(14, 'Tofu', Decimal('23.50'), "Mayumi's")
(15, 'Genen Shouyu', Decimal('15.50'), "Mayumi's")
(16, 'Pavlova', Decimal('17

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(38, 'Côte de Blaye', Decimal('263.50'), 'Aux joyeux ecclésiastiques')
(39, 'Chartreuse verte', Decimal('18.00'), 'Aux joyeux ecclésiastiques')
(35, 'Steeleye Stout', Decimal('18.00'), 'Bigfoot Breweries')
(67, 'Laughing Lumberjack Lager', Decimal('14.00'), 'Bigfoot Breweries')
(34, 'Sasquatch Ale', Decimal('14.00'), 'Bigfoot Breweries')
(11, 'Queso Cabrales', Decimal('21.00'), "Cooperativa de Quesos 'Las Cabras'")
(12, 'Queso Manchego La Pastora', Decimal('38.00'), "Cooperativa de Quesos 'Las Cabras'")
(58, 'Escargots de Bourgogne', Decimal('13.25'), 'Escargots Nouveaux')
(3, 'Aniseed Syrup', Decimal('10.00'), 'Exotic Liquids')
(1, 'Chai', Decimal('18.00'), 'Exotic Liquids')
(2, 'Chang', Decimal('19.00'), 'Exotic Liquids')
(62, 'Tarte au sucre', Decimal('49.30'), "Forêts d'érables")
(61, "Sirop d'érable", Decimal('28.50'), "Forêts d'érables")
(72, 'Mozzarella di Giovanni', Decimal('34.80'), 'Formaggi Fortini s.r.l.')
(31, 'Gorgonzola Telino', Decimal('12.50'), 'Formaggi Fortini s.r.l.')
(32, 'Mascarpone Fabioli', Decimal('32.00'), 'Formaggi Fortini s.r.l.')
(52, 'Filo Mix', Decimal('7.00'), "G'day, Mate")
(53, 'Perth Pasties', Decimal('32.80'), "G'day, Mate")
(51, 'Manjimup Dried Apples', Decimal('53.00'), "G'day, Mate")
(60, 'Camembert Pierrot', Decimal('34.00'), 'Gai pâturage')
(59, 'Raclette Courdavault', Decimal('55.00'), 'Gai pâturage')
(8, 'Northwoods Cranberry Sauce', Decimal('40.00'), "Grandma Kelly's Homestead")
(6, "Grandma's Boysenberry Spread", Decimal('25.00'), "Grandma Kelly's Homestead")
(7, "Uncle Bob's Organic Dried Pears", Decimal('30.00'), "Grandma Kelly's Homestead")
(25, 'NuNuCa Nuß-Nougat-Creme', Decimal('14.00'), 'Heli Süßwaren GmbH &amp; Co. KG')
(26, 'Gumbär Gummibärchen', Decimal('31.23'), 'Heli Süßwaren GmbH &amp; Co. KG')
(27, 'Schoggi Schokolade', Decimal('43.90'), 'Heli Süßwaren GmbH &amp; Co. KG')
(50, 'Valkoinen suklaa', Decimal('16.25'), 'Karkki Oy')
(76, 'Lakkalikööri', Decimal('18.00'), 'Karkki Oy')
(49, 'Maxilaku', Decimal('20.00'), 'Karkki Oy')
(42, 'Singaporean Hokkien Fried Mee', Decimal('14.00'), 'Leka Trading')
(43, 'Ipoh Coffee', Decimal('46.00'), 'Leka Trading')
(44, 'Gula Malacca', Decimal('19.45'), 'Leka Trading')
(45, 'Rogede sild', Decimal('9.50'), 'Lyngbysild')
(46, 'Spegesild', Decimal('12.00'), 'Lyngbysild')
(55, 'Pâté chinois', Decimal('24.00'), 'Ma Maison')
(54, 'Tourtière', Decimal('7.45'), 'Ma Maison')
(15, 'Genen Shouyu', Decimal('15.50'), "Mayumi's")
(13, 'Konbu', Decimal('6.00'), "Mayumi's")
(14, 'Tofu', Decimal('23.25'), "Mayumi's")
(40, 'Boston Crab Meat', Decimal('18.40'), 'New England Seafood Cannery')
(41, "Jack's New England Clam Chowder", Decimal('9.65'), 'New England Seafood Cannery')
(5, "Chef Anton's Gumbo Mix", Decimal('21.35'), 'New Orleans Ca6/Delights')
(65, 'Louisiana Fiery Hot Pepper Sauce', Decimal('21.05'), 'New Orleans Ca6/Delights')
(4, "Chef Anton's Ca6/Seasoning", Decimal('22.00'), 'New Orleans Ca6/Delights')
(66, 'Louisiana Hot Spiced Okra', Decimal('17.00'), 'New Orleans Ca6/Delights')
(30, 'Nord-Ost Matjeshering', Decimal('25.89'), 'Nord-Ost-Fisch Handelsgesellschaft mbH')
(69, 'Gudbrandsdalsost', Decimal('36.00'), 'Norske Meierier')
(33, 'Geitost', Decimal('2.50'), 'Norske Meierier')
(71, 'Flotemysost', Decimal('21.50'), 'Norske Meierier')
(57, 'Ravioli Angelo', Decimal('19.50'), 'Pasta Buttini s.r.l.')
(56, 'Gnocchi di nonna Alice', Decimal('38.00'), 'Pasta Buttini s.r.l.')
(18, 'Carnarvon Tigers', Decimal('62.50'), 'Pavlova, Ltd.')
(70, 'Outback Lager', Decimal('15.00'), 'Pavlova, Ltd.')
(16, 'Pavlova', Decimal('17.45'), 'Pavlova, Ltd.')
(63, 'Vegie-spread', Decimal('43.90'), 'Pavlova, Ltd.')
(17, 'Alice Mutton', Decimal('39.00'), 'Pavlova, Ltd.')
(23, 'Tunnbröd', Decimal('9.00'), 'PB Knäckebröd AB')
(22, "Gustaf's Knäckebröd", Decimal('21.00'), 'PB Knäckebröd AB')
(77, 'Original Frankfurter grüne Soße', Decimal('13.00'), 'Plutzer Lebensmittelgroßmärkte AG')
(28, 'Rössle Sauerkraut', Decimal('45.60'), 'Plutzer Lebensmittelgroßmärkte AG')
(75, 'Rhönbräu Klosterbier', Decimal('7.75'), 'Plutzer Lebensmittelgroßmärkte AG')
(29, 'Thüringer Rostbratwurst', Decimal('123.79'), 'Plutzer Lebensmittelgroßmärkte AG')
(64, 'Wimmers gute Semmelknödel', Decimal('33.25'), 'Plutzer Lebensmittelgroßmärkte AG')
(24, 'Guaraná Fantástica', Decimal('4.50'), 'Refrescos Americanas LTDA')
(20, "Sir Rodney's Marmalade", Decimal('81.00'), 'Specialty Biscuits, Ltd.')
(21, "Sir Rodney's Scones", Decimal('10.00'), 'Specialty Biscuits, Ltd.')
(68, 'Scottish Longbreads', Decimal('12.50'), 'Specialty Biscuits, Ltd.')
(19, 'Teatime Chocolate Biscuits', Decimal('9.20'), 'Specialty Biscuits, Ltd.')
(37, 'Gravad lax', Decimal('26.00'), 'Svensk Sjöföda AB')
(36, 'Inlagd Sill', Decimal('19.00'), 'Svensk Sjöföda AB')
(73, 'Röd Kaviar', Decimal('15.00'), 'Svensk Sjöföda AB')
(10, 'Ikura', Decimal('31.00'), 'Tokyo Traders')
(74, 'Longlife Tofu', Decimal('10.00'), 'Tokyo Traders')
(9, 'Mishi Kobe Niku', Decimal('97.00'), 'Tokyo Traders')
(47, 'Zaanse koeken', Decimal('9.50'), 'Zaanse Snoepfabriek')
(48, 'Chocolade', Decimal('12.75'), 'Zaanse Snoepfabriek')
(78, 'Stroopwafels', Decimal('9.75'), 'Zaanse Snoepfabriek')
</pre>
  
</details>

In [18]:
# Câu 2: Liệt danh sách các sản phẩm được đặt hàng bởi 'Thomas Hardy'
# Thông tin gồm: Order_Id, OrderDate, ProductName, Quantity, SalePrice và CustomerName
# Trong đó SalePrice = UnitPrice * Quantity

sql_cmd2 = '''
SELECT o.Id as OrderId, o.OrderDate, p.ProductName, oi.Quantity, 
       (oi.Quantity * oi.UnitPrice) AS SalePrice, 
       CONCAT(c.FirstName, ' ', c.LastName) AS CustomerName
FROM `customers` c JOIN `orders` o ON c.Id = o.CustomerId
                JOIN `orderitems` oi ON o.Id = oi.OrderId 
                JOIN `products` p ON oi.ProductId = p.Id;
'''

results_2 = run_sql(sql_cmd2)

for row in results_2:
    print(row)

(1, datetime.datetime(2012, 7, 4, 0, 0), 'Queso Cabrales', 12, Decimal('168.00'), 'Paul Henriot')
(1, datetime.datetime(2012, 7, 4, 0, 0), 'Singaporean Hokkien Fried Mee', 10, Decimal('98.00'), 'Paul Henriot')
(1, datetime.datetime(2012, 7, 4, 0, 0), 'Mozzarella di Giovanni', 5, Decimal('174.00'), 'Paul Henriot')
(2, datetime.datetime(2012, 7, 5, 0, 0), 'Tofu', 9, Decimal('167.40'), 'Karin Josephs')
(2, datetime.datetime(2012, 7, 5, 0, 0), 'Manjimup Dried Apples', 40, Decimal('1696.00'), 'Karin Josephs')
(3, datetime.datetime(2012, 7, 8, 0, 0), "Jack's New England Clam Chowder", 10, Decimal('77.00'), 'Mario Pontes')
(3, datetime.datetime(2012, 7, 8, 0, 0), 'Manjimup Dried Apples', 35, Decimal('1484.00'), 'Mario Pontes')
(3, datetime.datetime(2012, 7, 8, 0, 0), 'Louisiana Fiery Hot Pepper Sauce', 15, Decimal('252.00'), 'Mario Pontes')
(4, datetime.datetime(2012, 7, 8, 0, 0), "Gustaf's Knäckebröd", 6, Decimal('100.80'), 'Mary Saveley')
(4, datetime.datetime(2012, 7, 8, 0, 0), 'Ravioli An

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(108, datetime.datetime(2012, 11, 15, 0, 0), 'Guaraná Fantástica', 25, Decimal('3.60'), Decimal('90.00'), 'Thomas Hardy')
(108, datetime.datetime(2012, 11, 15, 0, 0), 'Ravioli Angelo', 25, Decimal('15.60'), Decimal('390.00'), 'Thomas Hardy')
(136, datetime.datetime(2012, 12, 16, 0, 0), 'Konbu', 20, Decimal('4.80'), Decimal('96.00'), 'Thomas Hardy')
(136, datetime.datetime(2012, 12, 16, 0, 0), 'Valkoinen suklaa', 15, Decimal('13.00'), Decimal('195.00'), 'Thomas Hardy')
(136, datetime.datetime(2012, 12, 16, 0, 0), 'Gnocchi di nonna Alice', 20, Decimal('30.40'), Decimal('608.00'), 'Thomas Hardy')
(206, datetime.datetime(2013, 2, 21, 0, 0), 'Chocolade', 15, Decimal('10.20'), Decimal('153.00'), 'Thomas Hardy')
(206, datetime.datetime(2013, 2, 21, 0, 0), 'Outback Lager', 25, Decimal('12.00'), Decimal('300.00'), 'Thomas Hardy')
(311, datetime.datetime(2013, 6, 4, 0, 0), 'Zaanse koeken', 25, Decimal('9.50'), Decimal('237.50'), 'Thomas Hardy')
(311, datetime.datetime(2013, 6, 4, 0, 0), 'Manjimup Dried Apples', 20, Decimal('53.00'), Decimal('1060.00'), 'Thomas Hardy')
(311, datetime.datetime(2013, 6, 4, 0, 0), 'Filo Mix', 30, Decimal('7.00'), Decimal('210.00'), 'Thomas Hardy')
(311, datetime.datetime(2013, 6, 4, 0, 0), 'Perth Pasties', 18, Decimal('32.80'), Decimal('590.40'), 'Thomas Hardy')
(311, datetime.datetime(2013, 6, 4, 0, 0), 'Röd Kaviar', 3, Decimal('15.00'), Decimal('45.00'), 'Thomas Hardy')
(460, datetime.datetime(2013, 10, 16, 0, 0), 'Pâté chinois', 21, Decimal('24.00'), Decimal('504.00'), 'Thomas Hardy')
(460, datetime.datetime(2013, 10, 16, 0, 0), 'Ravioli Angelo', 40, Decimal('19.50'), Decimal('780.00'), 'Thomas Hardy')
(460, datetime.datetime(2013, 10, 16, 0, 0), 'Outback Lager', 28, Decimal('15.00'), Decimal('420.00'), 'Thomas Hardy')
(494, datetime.datetime(2013, 11, 14, 0, 0), 'Chang', 15, Decimal('19.00'), Decimal('285.00'), 'Thomas Hardy')
(496, datetime.datetime(2013, 11, 17, 0, 0), 'Spegesild', 28, Decimal('12.00'), Decimal('336.00'), 'Thomas Hardy')
(521, datetime.datetime(2013, 12, 8, 0, 0), "Gustaf's Knäckebröd", 4, Decimal('21.00'), Decimal('84.00'), 'Thomas Hardy')
(521, datetime.datetime(2013, 12, 8, 0, 0), 'Gorgonzola Telino', 50, Decimal('12.50'), Decimal('625.00'), 'Thomas Hardy')
(521, datetime.datetime(2013, 12, 8, 0, 0), 'Camembert Pierrot', 15, Decimal('34.00'), Decimal('510.00'), 'Thomas Hardy')
(521, datetime.datetime(2013, 12, 8, 0, 0), 'Flotemysost', 12, Decimal('21.50'), Decimal('258.00'), 'Thomas Hardy')
(546, datetime.datetime(2013, 12, 24, 0, 0), "Jack's New England Clam Chowder", 14, Decimal('9.65'), Decimal('135.10'), 'Thomas Hardy')
(546, datetime.datetime(2013, 12, 24, 0, 0), 'Filo Mix', 8, Decimal('7.00'), Decimal('56.00'), 'Thomas Hardy')
(617, datetime.datetime(2014, 2, 2, 0, 0), 'Steeleye Stout', 4, Decimal('18.00'), Decimal('72.00'), 'Thomas Hardy')
(617, datetime.datetime(2014, 2, 2, 0, 0), 'Laughing Lumberjack Lager', 15, Decimal('14.00'), Decimal('210.00'), 'Thomas Hardy')
(673, datetime.datetime(2014, 3, 3, 0, 0), 'Valkoinen suklaa', 24, Decimal('16.25'), Decimal('390.00'), 'Thomas Hardy')
(706, datetime.datetime(2014, 3, 16, 0, 0), "Sir Rodney's Marmalade", 50, Decimal('81.00'), Decimal('4050.00'), 'Thomas Hardy')
(706, datetime.datetime(2014, 3, 16, 0, 0), 'Gorgonzola Telino', 50, Decimal('12.50'), Decimal('625.00'), 'Thomas Hardy')
(769, datetime.datetime(2014, 4, 10, 0, 0), 'Gorgonzola Telino', 15, Decimal('12.50'), Decimal('187.50'), 'Thomas Hardy')
(769, datetime.datetime(2014, 4, 10, 0, 0), 'Inlagd Sill', 16, Decimal('19.00'), Decimal('304.00'), 'Thomas Hardy')
</pre>
  
</details>

In [20]:
# Câu 3: Liệt kê danh sách các khách hàng chưa có đơn đặt hàng nào
sql_cmd3 = '''
SELECT c.Id, c.FirstName, c.LastName, c.City, c.Country, c.Phone
FROM `customers` c  LEFT JOIN `orders` o ON c.Id = o.CustomerId
WHERE o.CustomerId IS NULL;
'''

results_3 = run_sql(sql_cmd3)

for row in results_3:
    print(row)

(22, 'Diego', 'Roel', 'Madrid', 'Spain', '(91) 555 94 44')
(57, 'Marie', 'Bertrand', 'Paris', 'France', '(1) 42.34.22.66')


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(22, 'Diego', 'Roel', 'Madrid', 'Spain', '(91) 555 94 44')
(57, 'Marie', 'Bertrand', 'Paris', 'France', '(1) 42.34.22.66')
</pre>
  
</details>

In [22]:
# Câu 4: Liệt kê danh sách các nhà cung cấp có sản phẩm ngưng bán
sql_cmd4 = '''
SELECT s.Id, s.CompanyName, s.ContactName, s.ContactTitle, s.City, s.Country, s.Phone, s.Fax
FROM `products` p JOIN `suppliers` s ON p.SupplierId = s.Id
WHERE p.IsDiscontinued=1;
'''

results_4 = run_sql(sql_cmd4)

for row in results_4:
    print(row)

(2, 'New Orleans Ca6/Delights', 'Shelley Burke', None, 'New Orleans', 'USA', '(100) 555-4822', None)
(4, 'Tokyo Traders', 'Yoshi Nagase', None, 'Tokyo', 'Japan', '(03) 3555-5011', None)
(7, 'Pavlova, Ltd.', 'Ian Devling', None, 'Melbourne', 'Australia', '(03) 444-2343', '(03) 444-6588')
(10, 'Refrescos Americanas LTDA', 'Carlos Diaz', None, 'Sao Paulo', 'Brazil', '(11) 555 4640', None)
(12, 'Plutzer Lebensmittelgroßmärkte AG', 'Martin Bein', None, 'Frankfurt', 'Germany', '(069) 992755', None)
(12, 'Plutzer Lebensmittelgroßmärkte AG', 'Martin Bein', None, 'Frankfurt', 'Germany', '(069) 992755', None)
(20, 'Leka Trading', 'Chandra Leka', None, 'Singapore', 'Singapore', '555-8787', None)
(24, "G'day, Mate", 'Wendy Mackenzie', None, 'Sydney', 'Australia', '(02) 555-5914', '(02) 555-4873')


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(2, 'New Orleans Ca6/Delights', 'Shelley Burke', None, 'New Orleans', 'USA', '(100) 555-4822', None)
(4, 'Tokyo Traders', 'Yoshi Nagase', None, 'Tokyo', 'Japan', '(03) 3555-5011', None)
(7, 'Pavlova, Ltd.', 'Ian Devling', None, 'Melbourne', 'Australia', '(03) 444-2343', '(03) 444-6588')
(10, 'Refrescos Americanas LTDA', 'Carlos Diaz', None, 'Sao Paulo', 'Brazil', '(11) 555 4640', None)
(12, 'Plutzer Lebensmittelgroßmärkte AG', 'Martin Bein', None, 'Frankfurt', 'Germany', '(069) 992755', None)
(20, 'Leka Trading', 'Chandra Leka', None, 'Singapore', 'Singapore', '555-8787', None)
(24, "G'day, Mate", 'Wendy Mackenzie', None, 'Sydney', 'Australia', '(02) 555-5914', '(02) 555-4873')
</pre>
  
</details>

In [24]:
# Câu 5: Liệt kê danh sách các khách hàng chưa có đơn đặt hàng nào và các nhà cung cấp có sản phẩm ngưng bán
# Thông tin gồm: Type, Id, Name, City, Country, Phone
# Trong đó: 
#          + Type: cho biết đó là Customer hay Supplier
#          + Id: hiển thị Id của Customer hoặc Id của Supplier tùy theo Type
#          + Name: hiển thị CustomerName hoặc CompanyName tùy theo Type

sql_cmd5 = '''
(SELECT 'Customer' as Type, 
 c.Id, CONCAT(c.FirstName, ' ', c.LastName) as Name, c.City, c.Country, c.Phone
FROM `customers` c  LEFT JOIN `orders` o ON c.Id = o.CustomerId
WHERE o.CustomerId IS NULL)
UNION
(SELECT 'Supplier' as Type, 
 s.Id, s.CompanyName as Name, s.City, s.Country, s.Phone
FROM `products` p JOIN `suppliers` s ON p.SupplierId = s.Id
WHERE p.IsDiscontinued=1);
'''

results_5 = run_sql(sql_cmd5)

for row in results_5:
    print(row)

('Customer', 22, 'Diego Roel', 'Madrid', 'Spain', '(91) 555 94 44')
('Customer', 57, 'Marie Bertrand', 'Paris', 'France', '(1) 42.34.22.66')
('Supplier', 2, 'New Orleans Ca6/Delights', 'New Orleans', 'USA', '(100) 555-4822')
('Supplier', 4, 'Tokyo Traders', 'Tokyo', 'Japan', '(03) 3555-5011')
('Supplier', 7, 'Pavlova, Ltd.', 'Melbourne', 'Australia', '(03) 444-2343')
('Supplier', 10, 'Refrescos Americanas LTDA', 'Sao Paulo', 'Brazil', '(11) 555 4640')
('Supplier', 12, 'Plutzer Lebensmittelgroßmärkte AG', 'Frankfurt', 'Germany', '(069) 992755')
('Supplier', 20, 'Leka Trading', 'Singapore', 'Singapore', '555-8787')
('Supplier', 24, "G'day, Mate", 'Sydney', 'Australia', '(02) 555-5914')


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>('customer', 22, 'Diego Roel', 'Madrid', 'Spain', '(91) 555 94 44')
('customer', 57, 'Marie Bertrand', 'Paris', 'France', '(1) 42.34.22.66')
('Supplier', 2, 'New Orleans Ca6/Delights', 'New Orleans', 'USA', '(100) 555-4822')
('Supplier', 4, 'Tokyo Traders', 'Tokyo', 'Japan', '(03) 3555-5011')
('Supplier', 7, 'Pavlova, Ltd.', 'Melbourne', 'Australia', '(03) 444-2343')
('Supplier', 10, 'Refrescos Americanas LTDA', 'Sao Paulo', 'Brazil', '(11) 555 4640')
('Supplier', 12, 'Plutzer Lebensmittelgroßmärkte AG', 'Frankfurt', 'Germany', '(069) 992755')
('Supplier', 20, 'Leka Trading', 'Singapore', 'Singapore', '555-8787')
('Supplier', 24, "G'day, Mate", 'Sydney', 'Australia', '(02) 555-5914')
</pre>
  
</details>

In [27]:
# Câu 6: Liệt kê thông tin đặt hàng của các khách hàng (kể cả các khách hàng chưa có đặt hàng), chỉ liệt kê 10 dòng đầu tiên.
sql_cmd6 = '''
SELECT CONCAT(c.FirstName, ' ', c.LastName) AS NAME,
	   o.Id as OrderId, o.OrderDate
FROM `orders` o RIGHT JOIN `customers` c ON o.CustomerId=c.Id
ORDER BY o.Id
LIMIT 10;
'''

results_6 = run_sql(sql_cmd6)

for row in results_6:
    print(row)

('Marie Bertrand', None, None)
('Diego Roel', None, None)
('Paul Henriot', 1, datetime.datetime(2012, 7, 4, 0, 0))
('Karin Josephs', 2, datetime.datetime(2012, 7, 5, 0, 0))
('Mario Pontes', 3, datetime.datetime(2012, 7, 8, 0, 0))
('Mary Saveley', 4, datetime.datetime(2012, 7, 8, 0, 0))
('Pascale Cartrain', 5, datetime.datetime(2012, 7, 9, 0, 0))
('Mario Pontes', 6, datetime.datetime(2012, 7, 10, 0, 0))
('Yang Wang', 7, datetime.datetime(2012, 7, 11, 0, 0))
('Michael Holz', 8, datetime.datetime(2012, 7, 12, 0, 0))


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>('Marie Bertrand', None, None)
('Diego Roel', None, None)
('Paul Henriot', 1, datetime.datetime(2012, 7, 4, 0, 0))
('Karin Josephs', 2, datetime.datetime(2012, 7, 5, 0, 0))
('Mario Pontes', 3, datetime.datetime(2012, 7, 8, 0, 0))
('Mary Saveley', 4, datetime.datetime(2012, 7, 8, 0, 0))
('Pascale Cartrain', 5, datetime.datetime(2012, 7, 9, 0, 0))
('Mario Pontes', 6, datetime.datetime(2012, 7, 10, 0, 0))
('Yang Wang', 7, datetime.datetime(2012, 7, 11, 0, 0))
('Michael Holz', 8, datetime.datetime(2012, 7, 12, 0, 0))
</pre>
  
</details>

In [29]:
conn.close()